torch/csrc/api/src/nn/module.cpp

```diff
@@ -0,0 +1,125 @@
+#include <torch/nn/module.h>
+#include <torch/detail/ordered_dict.h>
+#include <torch/nn/cursor.h>
+
+#include <torch/csrc/autograd/variable.h>
+
+#include <ATen/ATen.h>
+
+#include <memory>
+#include <string>
+#include <unordered_map>
+#include <vector>
+
+namespace torch { namespace nn {
+
+Module::Module(const char* name) : name_(name), is_training_(false) {}
+
+Module::~Module() = default;
+
+std::unique_ptr<Module> Module::clone() {
+  TORCH_ERROR(
+      "clone() has not been implemented for %s. "
+      "Use the copy constructor if you don't require polymorphic cloning. "
+      "Otherwise, subclass CloneableModule<%s> to inherit cloning behavior.",
+      name_,
+      name_);
+}
+
+std::vector<Tensor> Module::operator()(const std::vector<Tensor>& inputs) {
+  return forward(inputs);
+}
+
+// Train/Eval mode
+void Module::train() {
+  is_training_ = true;
+}
+
+void Module::eval() {
+  is_training_ = false;
+}
+
+bool Module::is_training() const noexcept {
+  return is_training_;
+}
+
+// Recursive Transformations
+void Module::cpu() {
+  // parameters().apply([](Tensor& tensor) { tensor.toBackend_(kCPU); });
+}
+
+void Module::cuda() {
+  // parameters().apply([](Tensor& tensor) { tensor.toBackend_(kCUDA); });
+}
+
+void Module::type(at::ScalarType new_type) {
+  // parameters().apply([=](Tensor& tensor) { tensor.toType_(new_type); });
+  // buffers().apply([=](Tensor& tensor) { tensor.toType_(new_type); });
```

Comment:
These are implemented now, right?

Reply:
I'm waiting for the "Autograd API" in Tensor. I need the `set_type` behavior, which is currently `temporary_hack_set_type` and will be `detail::set_data` in the autograd API PR

Reply:
Cool, thanks for the explanation!